MapDB Persistence: PercentType restore needs to allow Decimal values (Issue #3437)

mrguessed opened the pull request:
openHAB Issue #3437

In openHAB, `PercentType` is derived from `DecimalType`.  The
MapDB Persistence is pre-parsing the value, assuming it's an `Integer`,
and failing when it wasn't an Integer.

In some cases, this would cause the openHAB startup process to enter a never ending loop, since the Persistence-Restore functionality was failing.

With this change, we simply hand the String value to `PercentType.valueOf()` and let it perform any necessary conversions.

Signed-off-by: Mark Clark <mr.guessed@gmail.com>

mrguessed commented:
@JSurf,
You you mind taking a first pass review?

lewie commented:
@mrguessed,
try to use https://github.com/openhab/openhab/pull/3248 instead. The problem is the conversation between 'double' and 'BigDecimal'.
More Infos:  https://github.com/openhab/openhab/wiki/JDBC-Persistence#number-precision.

Rgds
Helmut

mrguessed commented:
@lewie,
Actually I do want MapDB, and don't need the extra components that JDBC would introduce (I already use RRD for my numeric values, and don't require history otherwise)

This PR is specifically to address the data handling in the MapDB persistence binding, where it (externally) attempts to convert to an int, where the input value needn't necessarily be an int. (See the diffs, from my fix above, for reference)

lewie commented:
@mrguessed,
you are right!
Missed some details. 

Thanks for the clarification
Helmut

JSurf commented:
Pull request looks good to me. Perhaps für consistency DecimalType And HSBType then should use valueOf aswell ? At least JPA and mysql binding also use the same code is it also a problem there ?

mrguessed commented:
@JSurf,
I converted the remaining types to use `*Type.valueOf(...)` instead of `new *Type(...)` and updated the PR's Branch. 

`DecimalType` was technically the only one that warranted this conversion, but the others were done for consistency (even though they internally just `return new *Type(...)`

If JPA and/or mysql handle `PercentType` in the way MapDB Persistence did, then there's a potential problem lurking in that code if they're not scrubbing the value upon save (I haven't looked).  Did you file a ticket for them and/or open a forum discussion for them (I don't use either of these Bindings)

The other changes were cosmetic, and not an issue.

teichsta commented:
Thanks @mrguessed!

@lewie could you please have a look at the JPA binding and follow-up on @JSurfs comment?

Thanks, Thomas E.-E.
